### PR TITLE
Add eth_getBlockReceipts() JSON/RPC method

### PR DESCRIPTION
**Pull request by matthew1001**
## PR description
Adds `eth_getBlockReceipts` JSON/RPC method as described by Ethereum API specification https://github.com/ethereum/execution-apis/pull/438

**Notes for PR reviewer & docs**:

- The specification says that the correct response for a block that cannot be found is an empty array, e.g. https://github.com/ethereum/execution-apis/blob/main/tests/eth_getBlockReceipts/get-block-receipts-future.io Besu  appears to have a convention of returning a `-32000` error code with message `Block not found` for such cases. I have done the same with this new JSON/RPC method but welcome any thoughts on this from the PR reviewer.
- I don't think this query requires any changes/additions to graphql. The current schema covers the basic EVM objects (block, transaction, log etc) and is already designed to provide an alternative way of doing what this new JSON/RPC method adds (getting all receipts for a block in a single call) so I haven't made any graphql changes.

## Fixed Issue(s)
Closes https://github.com/hyperledger/besu/issues/5751

### CODE REVIEW

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java:1]: also have a look at EthJsonRpcHttpBySpecTest - maybe add some "by spec" tests

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    final BlockReceiptsResult result =
+        (BlockReceiptsResult) ((JsonRpcSuccessResponse) actualResponse).getResult();
+
+    assertThat(result.getResults().size()).isEqualTo(2);
+
+    // Check TX1 receipt is correct
+    TransactionReceiptResult tx1 = result.getResults().get(0);
+    assertThat(tx1.getBlockNumber()).isEqualTo("0x1");
+    assertThat(tx1.getEffectiveGasPrice()).isEqualTo("0x4a7ebf2e");
+    assertThat(tx1.getTo()).isEqualTo("0x6ada2e11049e5fc54fcdf2a97996d9b2aa80fe71");
```
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java:150]: final Transaction expectedTx1 = blockchain.getBlockByNumber(BLOCKCHAIN_LENGTH - 1).get().getBody().getTransactions().get(0);

    // Check TX1 receipt is correct
    TransactionReceiptResult tx1 = result.getResults().get(0);
    assertThat(tx1.getBlockNumber()).isEqualTo("0x"+(BLOCKCHAIN_LENGTH-1));
    assertThat(tx1.getEffectiveGasPrice()).isNotEmpty();
    assertThat(tx1.getType()).isEqualTo(expectedTx1.getType());
    assertThat(tx1.getTo()).isEqualTo(expectedTx1.getTo().get().toString());

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    TransactionReceiptResult tx2 = result.getResults().get(1);
+    assertThat(tx2.getBlockNumber()).isEqualTo("0x4");
+    assertThat(tx2.getEffectiveGasPrice()).isEqualTo("0x9f7cd42");
+    assertThat(tx2.getTo()).isEqualTo("0x46530778ec4a61cbea38698d2efc0c618ea0641f");
+    assertThat(tx2.getType()).isEqualTo("0x2");
+  }
+
+  @Test
+  public void twoReceiptsForBlockOne() {
+    /* Block generator defaults to 2 transactions per mocked block */
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java:138]:
```suggestion
    /* Block generator defaults to 2 transactions per block */
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (matthew1001) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java:150]: Yeah that's a good point. I'll make those suggested changes so the test is more reliable.

[the remark this answers is not in the document]
Comment (matthew1001) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java]: The latest commit updates the unit tests as you've suggested @macfarla. I added `getEthSerializedType()` to `TransactionType` so there is a single place in the codebase where `FRONTIER` transactions are mapped from their enum value of `0xf8` to their eth serialized value of `0x00`.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  @Test
+  public void blockNotFoundWhenHash63CharsLong() {
+    /* Valid hash with 63 chars in - should result in block not found */
+    JsonRpcResponse actualResponse = method.response(requestWithParams(HASH_63_CHARS_LONG));
+    assertThat(actualResponse).usingRecursiveComparison().isEqualTo(blockNotFoundResponse);
+  }
+
+  @Test
+  public void blockNotFoundForRandomHash() {
+    /* Valid random hash - should result in block not found (effectively impossible for it to be a valid block) */
```
[the remark this answers is not in the document]
Comment (matthew1001) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java]: Yeah fair point, I'll remove that test for now.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java
Comment (matthew1001) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetBlockReceiptsTest.java:1]: Thanks for the pointer. I'd found the graphql spec tests but not spotted general JSON/RPC spec tests. Latest commit adds a variety to cover by hash, by number, and by tag.

**Comment by matthew1001** [comment]
I think all of the latest comments are addressed @macfarla I'm away for a few days now, feel free to merge if you're happy or if you have more comments I'm happy to take a look when I'm back.

**Review by macfarla — APPROVED** [review]

**Pull request merged**